### PR TITLE
Detect WSL1 and error

### PR DESCRIPTION
**Pull request by cole-h**
WSL1 is not supported because some things Nix relies on to work are not available.

##### Description

Closes https://github.com/DeterminateSystems/nix-installer/issues/281.

##### Checklist

- [ ] Formatted with `cargo fmt`
- [ ] Built with `nix build`
- [ ] Ran flake checks with `nix flake check`
- [ ] Added or updated relevant tests (leave unchecked if not applicable)
- [ ] Added or updated relevant documentation (leave unchecked if not applicable)
- [ ] Linked to related issues (leave unchecked if not applicable)

##### Validating with `install.determinate.systems`

If a maintainer has added the `upload to s3` label to this PR, it will become available for installation via `install.determinate.systems`:

```shell
curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix/pr/$PR_NUMBER | sh -s -- install
```

### CODE REVIEW

File: src/planner/linux.rs [hunk 1]
```diff
@@ -39,6 +39,10 @@ impl Planner for Linux {
             return Err(PlannerError::NixOs);
         }
 
+        if std::env::var("WSL_DISTRO_NAME").is_ok() && !Path::new("/run/WSL").exists() {
```
Comment (cole-h) [inline src/planner/linux.rs:42]: Bad news: I was lied to that "`/run/WSL` existing means WSL2". Or at least, it doesn't determine if the _current_ WSL instance is, just that there is _some_ WSL2 instance available (I have both v1 and v2 instances, and /run/WSL exists in both).

But apparently only WSL2 has the `WSL_INTEROP` env var, so let's try that.

**Comment by cole-h** [comment]
Nice.

image

image

File: src/planner/mod.rs [hunk 2]
```diff
@@ -301,6 +301,8 @@ pub enum PlannerError {
     NixOs,
     #[error("`nix` is already a valid command, so it is installed")]
     NixExists,
+    #[error("WSL1 is not supported, please upgrade to WSL2")]
```
Comment (Hoverbear) [inline src/planner/mod.rs:304]: Think we can give them a link how to?

Comment (cole-h) [reply src/planner/mod.rs:304]:
```suggestion
    #[error("WSL1 is not supported, please upgrade to WSL2: https://learn.microsoft.com/en-us/windows/wsl/install#upgrade-version-from-wsl-1-to-wsl-2")]
```

Just gotta hope MS doesn't move that link...

**Review by Hoverbear — APPROVED** [review]

**Comment by stuv** [comment]
Sorry to open up this merge. But I wanted to install Determinate Systems Nix on WSL1 and I didn't get the error like this.

The one I bumped into was: 

```
Proceed? ([Y]es/[n]o/[e]xplain): y
 INFO Step: Create directory `/nix`
 INFO Step: Provision Nix
 INFO Step: Create build users (UID 30001-30032) and group (GID 30000)
 INFO Step: Configure Nix
ERROR
   0: Install failure
   1: Error executing action
   2: Action `configure_nix` errored
   3: Action `setup_default_profile` errored
   4: Failed to execute command with status 1 `HOME="/root" NIX_SSL_CERT_FILE="/nix/store/852a51xd3m7jyzifqimxc3jf3w2ipc4k-nss-cacert-3.101/etc/ssl/certs/ca-bundle.crt" "/nix/store/hdy82qidsybc3fg561pqfwagv44vschb-nix-2.24.10/bin/nix-env" "--option" "substitute" "false" "-i" "/nix/store/hdy82qidsybc3fg561pqfwagv44vschb-nix-2.24.10"`, stdout:
      stderr: installing 'nix-2.24.10'
      error:
             … while setting up the build environment

             error: unable to load seccomp BPF program: Invalid argument

   4:

Location:
   src/cli/subcommand/install.rs:256
```
   
This was all with the default settings for as far as I can remember. Using the standard install command from the website:
`curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix | \
  sh -s -- install`
  
I assume nix still isn't supported on WSL1? It's unfortunate, but with this computer I can't update it to WSL2, so I am stuck with WSL1 and would like to see if in any way it's possible to have some nix on this machine to make life a bit more bareable.

**Comment by cole-h** [comment]
You are correct in that WSL1 is still not (explicitly) supported. If you can provide information on a more reliable way of detecting if the WSL version used is WSL1 and not WSL2, I'd be grateful!

In the meantime, while this "bug" exists, you can see if using `--init none` and one (or both) of `--extra-conf "sandbox = false"` and `--extra-conf "filter-syscalls = false"` works to install Nix -- if so, then we might be able to make WSL1 "usable" (by documenting this combination of flags).

**Comment by stuv** [comment]
Sorry for the late late reply. But just reporting back that this worked.

I did have to do a recursive chown on /nix with $USER, because it would throw me this error:

> error: could not set permissions on '/nix/var/nix/profiles/per-user' to 755: Operation not permitted

Not sure if this was due to my earlier playing around with it and not knowing what I am doing. 
But am testing some thinks out now and it seems that nix is working on WSL1 :)

Thanks for the suggestions @cole-h !!

**Pull request merged**